src/charm.py

```diff
@@ -6,36 +6,57 @@
 """Charm Jenkins."""
 
 import logging
-from typing import TYPE_CHECKING, Any, Dict, cast
+import typing
 
-from ops.charm import CharmBase, PebbleReadyEvent
+from ops.charm import ActionEvent, CharmBase, PebbleReadyEvent, RelationJoinedEvent
 from ops.main import main
-from ops.model import ActiveStatus, BlockedStatus, MaintenanceStatus
+from ops.model import ActiveStatus, BlockedStatus, Container, MaintenanceStatus
 from ops.pebble import Layer
 
-from jenkins import JENKINS_WEB_URL, calculate_env, unlock_jenkins, wait_jenkins_ready
-from types_ import JenkinsEnvironmentMap
+import jenkins
+import state
 
-if TYPE_CHECKING:
+if typing.TYPE_CHECKING:
     from ops.pebble import LayerDict  # pragma: no cover
 
 
 logger = logging.getLogger(__name__)
 
 
+class AgentRelationData(typing.TypedDict):
+    """Relation data required for adding the Jenkins agent.
+
+    Attrs:
+        url: The Jenkins server url.
+        secret: The secret for agent node.
+    """
+
+    url: str
+    secret: str
+
+
 class JenkinsK8SOperatorCharm(CharmBase):
     """Charm Jenkins."""
 
-    def __init__(self, *args: Any):
+    def __init__(self, *args: typing.Any):
         """Initialize the charm and register event handlers.
 
         Args:
             args: Arguments to initialize the char base.
         """
         super().__init__(*args)
+        self.state = state.State.from_charm(self.model.config)
+
         self.framework.observe(self.on.jenkins_pebble_ready, self._on_jenkins_pebble_ready)
+        self.framework.observe(self.on["agent"].relation_joined, self._on_agent_relation_joined)
```

Comment:
Perhaps the observer pattern here?

Reply:
Refactored to agent observer, thank you!